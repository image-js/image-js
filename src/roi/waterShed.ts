import PriorityQueue from 'js-priority-queue';

import { RoiMapManager } from '..';
import { Image } from '../Image';
import { Mask } from '../Mask';
import getExtrema from '../compute/getExtrema';
import { Point } from '../geometry';
import { filterPoints } from '../utils/geometry/filterPoints';
import checkProcessable from '../utils/validators/checkProcessable';
/**
 * Point interface that is used in the queue data structure.
 */
interface PointWithIntensity {
  /**
   * @param row - Row of a point.
   */
  row: number;
  /**
   * @param column - Column of a point.
   */
  column: number;
  /**
   * @param intensity - Value of a point.
   */
  intensity: number;
}

interface WaterShedOptions {
  /**
   * @param points - Points which should be filled by watershed filter.
   * @default - minimum points from getExtrema() function.
   */
  points?: Point[];
  /**
   * @param mask - A binary image, the same size as the image. The algorithm will fill only if the current pixel in the binary mask is not null.
   * @default undefined
   */
  mask?: Mask;
  /**
   * @param threshold - Limit of filling. Maximum value that pixel can have.
   * @default 1
   */
  threshold?: number;
}
/**
 * This method allows to create a ROIMap using the water shed algorithm. By default this algorithm
 * will fill the holes and therefore the lowest value of the image (black zones).
 * If no points are given, the function will look for all the minimal points.
 * If no mask is given the algorithm will completely fill the image.
 * Please take care about the value that has be in the mask ! In order to be coherent with the expected mask,
 * meaning that if it is a dark zone, the mask will be dark the normal behavior to fill a zone
 * is that the mask pixel is clear (value of 0) !
 * If you are looking for 'maxima' the image must be inverted before applying the algorithm
 * @param image - Image that the filter will be applied to.
 * @param options - WaterShedOptions
 * @returns RoiMapManager
 */
export function waterShed(
  image: Image,
  options: WaterShedOptions,
): RoiMapManager {
  let { points } = options;
  const { mask, threshold = 1 } = options;
  const currentImage = image;
  checkProcessable(image, {
    bitDepth: [8, 16],
    components: 1,
  });

  const fillMaxValue = threshold * image.maxValue;

  // WaterShed is done from points in the image. We can either specify those points in options,
  // or it is gonna take the minimum locals of the image by default.
  if (!points) {
    points = getExtrema(image, {
      kind: 'minimum',
      mask,
    });
    points = filterPoints(points, image, { kind: 'minimum' });
  }

  const maskExpectedValue = 0;

  const data = new Int32Array(currentImage.size);
  const width = currentImage.width;
  const height = currentImage.height;
  const toProcess = new PriorityQueue({
    comparator: (a: PointWithIntensity, b: PointWithIntensity) =>
      a.intensity - b.intensity,
    strategy: PriorityQueue.BinaryHeapStrategy,
  });
  for (let i = 0; i < points.length; i++) {
    const index = points[i].column + points[i].row * width;
    data[index] = -i - 1;
    const intensity = currentImage.getValueByIndex(index, 0);
    if (intensity <= fillMaxValue) {
      toProcess.queue({
        column: points[i].column,
        row: points[i].row,
        intensity,
      });
    }
  }
  const dxs = [+1, 0, -1, 0, +1, +1, -1, -1];
  const dys = [0, +1, 0, -1, +1, -1, +1, -1];
  // Then we iterate through each points

  while (toProcess.length > 0) {
    const currentPoint = toProcess.dequeue();
    const currentValueIndex = currentPoint.column + currentPoint.row * width;
    for (let dir = 0; dir < 4; dir++) {
      const newX = currentPoint.column + dxs[dir];
      const newY = currentPoint.row + dys[dir];
      if (newX >= 0 && newY >= 0 && newX < width && newY < height) {
        const currentNeighbourIndex = newX + newY * width;
        if (
          !mask ||
          mask.getBitByIndex(currentNeighbourIndex) === maskExpectedValue
        ) {
          const intensity = currentImage.getValueByIndex(
            currentNeighbourIndex,
            0,
          );
          if (intensity <= fillMaxValue && data[currentNeighbourIndex] === 0) {
            data[currentNeighbourIndex] = data[currentValueIndex];
            toProcess.queue({
              column: currentPoint.column + dxs[dir],
              row: currentPoint.row + dys[dir],
              intensity,
            });
          }
        }
      }
    }
  }
  const nbNegative = points.length;
  const nbPositive = 0;

  return new RoiMapManager({
    data,
    nbPositive,
    nbNegative,
    width: image.width,
    height: image.height,
  });
}
