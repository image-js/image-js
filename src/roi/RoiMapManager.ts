import { Matrix } from 'ml-matrix';

import { Mask } from '../Mask';

import { Roi } from './Roi';
import { FromMaskOptions, fromMask } from './fromMask';
import { getRois, GetRoisOptions } from './getRois';

export interface RoiManager {
  getRois(options: GetRoisOptions): Roi[];
}

export interface RoiMap {
  /**
   * Width of the map.
   */
  width: number;
  /**
   * Height of the map.
   */
  height: number;
  /**
   * Data of the ROIs. Each ROI is associated with a negative or a positive value,
   * depending if it derives from a zone made of zeros or ones in the original mask.
   */
  data: Int16Array;
  /**
   * Number of distinct positive values in the ROI map.
   *
   */
  nbPositive: number;
  /**
   * Number of distinct negative values in the ROI map.
   *
   */
  nbNegative: number;
}
export class RoiMapManager implements RoiManager {
  private map: RoiMap;
  public whiteRois: Roi[];
  public blackRois: Roi[];

  public constructor(map: RoiMap) {
    this.map = map;
    this.whiteRois = [];
    this.blackRois = [];
  }

  /**
   * Return the ROI map of the RoiMapManager.
   * @returns - The ROI map.
   */
  public getMap(): RoiMap {
    return this.map;
  }

  /**
   * Return the value at the given coordinates in an ROI map.
   * @param column - Column of the value.
   * @param row - Row of the value.
   * @returns The value at the given coordinates.
   */
  public getMapValue(column: number, row: number) {
    return this.map.data[this.map.width * row + column];
  }

  /**
   * Returns the ROI map as a correct width and height matrix.
   * @returns The ROI map matrix.
   */
  public getMapMatrix(): number[][] {
    return Matrix.from1DArray(
      this.map.height,
      this.map.width,
      this.map.data,
    ).to2DArray();
  }

  public getRois(options: GetRoisOptions = {}): Roi[] {
    return getRois(this, options);
  }

  public getRoiById(roiID: number): Roi {
    const rois = this.getRois({ kind: 'bw' });
    const foundRoi = rois.find((roi) => roi.id === roiID);
    if (!foundRoi) {
      throw new Error(`invalid ID: ${roiID}`);
    }
    return foundRoi;
  }

  public static fromMask(mask: Mask, options: FromMaskOptions = {}) {
    return fromMask(mask, options);
  }
}
