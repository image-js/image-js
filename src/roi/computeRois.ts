import { Roi } from './Roi';
import { RoiMapManager } from './RoiMapManager';

/**
 * Generate an array of ROIs based on an ROI map manager.
 * @param roiMapManager - Roi map manager to use.
 */
export function computeRois(roiMapManager: RoiMapManager): void {
  const map = roiMapManager.getMap();

  const whites = new Array(map.nbPositive);
  const blacks = new Array(map.nbNegative);
  for (let i = 0; i < map.nbPositive; i++) {
    whites[i] = {
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: i + 1,
    };
  }

  for (let i = 0; i < map.nbNegative; i++) {
    blacks[i] = {
      borderLengths: [],
      borderIDs: [],
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: -i - 1,
    };
  }

  for (let row = 0; row < map.height; row++) {
    for (let column = 0; column < map.width; column++) {
      const currentIndex = roiMapManager.getMapValue(column, row);
      if (currentIndex === 0) {
        continue;
      }
      let currentRoi;
      if (currentIndex < 0) {
        currentRoi = blacks[-currentIndex - 1];
      } else {
        currentRoi = whites[currentIndex - 1];
      }

      currentRoi.surface++;

      if (row < currentRoi.minRow) {
        currentRoi.minRow = row;
      }
      if (row > currentRoi.maxRow) {
        currentRoi.maxRow = row;
      }
      if (column < currentRoi.minColumn) {
        currentRoi.minColumn = column;
      }
      if (column > currentRoi.maxColumn) {
        currentRoi.maxColumn = column;
      }
    }
  }

  roiMapManager.whiteRois = new Array<Roi>(map.nbPositive);
  roiMapManager.blackRois = new Array<Roi>(map.nbNegative);

  for (let i = 0; i < map.nbNegative; i++) {
    const width = blacks[i].maxColumn - blacks[i].minColumn + 1;
    const height = blacks[i].maxRow - blacks[i].minRow + 1;
    const origin = { row: blacks[i].minRow, column: blacks[i].minColumn };
    const id = blacks[i].id;
    const surface = blacks[i].surface;
    const blackRoi = new Roi(map, id, width, height, origin, surface);
    roiMapManager.blackRois[i] = blackRoi;
  }

  for (let i = 0; i < map.nbPositive; i++) {
    const width = whites[i].maxColumn - whites[i].minColumn + 1;
    const height = whites[i].maxRow - whites[i].minRow + 1;
    const origin = { row: whites[i].minRow, column: whites[i].minColumn };
    const id = whites[i].id;
    const surface = whites[i].surface;
    const whiteRoi = new Roi(map, id, width, height, origin, surface);

    roiMapManager.whiteRois[i] = whiteRoi;
  }
}
