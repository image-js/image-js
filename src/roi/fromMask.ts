import { Mask } from '..';
import { assert } from '../utils/validators/assert';

import { RoiMapManager } from './RoiMapManager';
import { maxNumberRois, maxRoiId } from './utils/constants';

export interface FromMaskOptions {
  /**
   * Consider pixels connected by corners as same ROI?
   * @default `false`
   */
  allowCorners?: boolean;
}

/**
 * Extract the ROIs of an image.
 * @param mask - Mask to extract the ROIs from.
 * @param options - From mask options.
 * @returns The corresponding ROI manager.
 */
export function fromMask(
  mask: Mask,
  options: FromMaskOptions = {},
): RoiMapManager {
  const { allowCorners = false } = options;

  const MAX_ARRAY = maxNumberRois - 1; // 65535 should be enough for most of the cases

  const maxPositiveId = maxRoiId - 1;
  const maxNegativeId = -maxRoiId;

  // based on a binary image we will create plenty of small images
  const data = new Int32Array(mask.size); // maxValue: maxPositiveId, minValue: maxNegativeId

  // split will always return an array of images
  let positiveId = 0;
  let negativeId = 0;

  const columnToProcess = new Uint16Array(maxNumberRois);
  const rowToProcess = new Uint16Array(maxNumberRois);

  for (let column = 0; column < mask.width; column++) {
    for (let row = 0; row < mask.height; row++) {
      if (data[row * mask.width + column] === 0) {
        // need to process the whole surface
        analyseSurface(column, row);
      }
    }
  }
  // x column
  // y row
  function analyseSurface(column: number, row: number) {
    let from = 0;
    let to = 0;
    const targetState = mask.getBit(column, row);
    const id = targetState ? ++positiveId : --negativeId;
    assert(
      positiveId <= maxPositiveId && negativeId >= maxNegativeId,
      'too many regions of interest',
    );
    columnToProcess[0] = column;
    rowToProcess[0] = row;
    while (from <= to) {
      const currentColumn = columnToProcess[from & MAX_ARRAY];
      const currentRow = rowToProcess[from & MAX_ARRAY];
      data[currentRow * mask.width + currentColumn] = id;
      // need to check all around mask pixel
      if (
        currentColumn > 0 &&
        data[currentRow * mask.width + currentColumn - 1] === 0 &&
        mask.getBit(currentColumn - 1, currentRow) === targetState
      ) {
        // LEFT
        to++;
        columnToProcess[to & MAX_ARRAY] = currentColumn - 1;
        rowToProcess[to & MAX_ARRAY] = currentRow;
        data[currentRow * mask.width + currentColumn - 1] = maxNegativeId;
      }
      if (
        currentRow > 0 &&
        data[(currentRow - 1) * mask.width + currentColumn] === 0 &&
        mask.getBit(currentColumn, currentRow - 1) === targetState
      ) {
        // TOP
        to++;
        columnToProcess[to & MAX_ARRAY] = currentColumn;
        rowToProcess[to & MAX_ARRAY] = currentRow - 1;
        data[(currentRow - 1) * mask.width + currentColumn] = maxNegativeId;
      }
      if (
        currentColumn < mask.width - 1 &&
        data[currentRow * mask.width + currentColumn + 1] === 0 &&
        mask.getBit(currentColumn + 1, currentRow) === targetState
      ) {
        // RIGHT
        to++;
        columnToProcess[to & MAX_ARRAY] = currentColumn + 1;
        rowToProcess[to & MAX_ARRAY] = currentRow;
        data[currentRow * mask.width + currentColumn + 1] = maxNegativeId;
      }
      if (
        currentRow < mask.height - 1 &&
        data[(currentRow + 1) * mask.width + currentColumn] === 0 &&
        mask.getBit(currentColumn, currentRow + 1) === targetState
      ) {
        // BOTTOM
        to++;
        columnToProcess[to & MAX_ARRAY] = currentColumn;
        rowToProcess[to & MAX_ARRAY] = currentRow + 1;
        data[(currentRow + 1) * mask.width + currentColumn] = maxNegativeId;
      }
      if (allowCorners) {
        if (
          currentColumn > 0 &&
          currentRow > 0 &&
          data[(currentRow - 1) * mask.width + currentColumn - 1] === 0 &&
          mask.getBit(currentColumn - 1, currentRow - 1) === targetState
        ) {
          // TOP LEFT
          to++;
          columnToProcess[to & MAX_ARRAY] = currentColumn - 1;
          rowToProcess[to & MAX_ARRAY] = currentRow - 1;
          data[(currentRow - 1) * mask.width + currentColumn - 1] =
            maxNegativeId;
        }
        if (
          currentColumn < mask.width - 1 &&
          currentRow > 0 &&
          data[(currentRow - 1) * mask.width + currentColumn + 1] === 0 &&
          mask.getBit(currentColumn + 1, currentRow - 1) === targetState
        ) {
          // TOP RIGHT
          to++;
          columnToProcess[to & MAX_ARRAY] = currentColumn + 1;
          rowToProcess[to & MAX_ARRAY] = currentRow - 1;
          data[(currentRow - 1) * mask.width + currentColumn + 1] =
            maxNegativeId;
        }
        if (
          currentColumn > 0 &&
          currentRow < mask.height - 1 &&
          data[(currentRow + 1) * mask.width + currentColumn - 1] === 0 &&
          mask.getBit(currentColumn - 1, currentRow + 1) === targetState
        ) {
          // BOTTOM LEFT
          to++;
          columnToProcess[to & MAX_ARRAY] = currentColumn - 1;
          rowToProcess[to & MAX_ARRAY] = currentRow + 1;
          data[(currentRow + 1) * mask.width + currentColumn - 1] =
            maxNegativeId;
        }
        if (
          currentColumn < mask.width - 1 &&
          currentRow < mask.height - 1 &&
          data[(currentRow + 1) * mask.width + currentColumn + 1] === 0 &&
          mask.getBit(currentColumn + 1, currentRow + 1) === targetState
        ) {
          // BOTTOM RIGHT
          to++;
          columnToProcess[to & MAX_ARRAY] = currentColumn + 1;
          rowToProcess[to & MAX_ARRAY] = currentRow + 1;
          data[(currentRow + 1) * mask.width + currentColumn + 1] =
            maxNegativeId;
        }
      }

      from++;

      assert(
        to - from <= MAX_ARRAY,
        'fromMask can not finish, the array to manage internal data is not big enough.' +
          'You could improve mask by changing MAX_ARRAY',
      );
    }
  }
  return new RoiMapManager({
    width: mask.width,
    height: mask.height,
    data,
    nbNegative: Math.abs(negativeId),
    nbPositive: positiveId,
  });
}
