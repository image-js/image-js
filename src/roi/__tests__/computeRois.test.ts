import { expect, test } from 'vitest';

import { RoiMapManager } from '../RoiMapManager.js';
import { computeRois } from '../computeRois.js';
import { fromMask } from '../fromMask.js';
import { waterShed } from '../waterShed.js';

test('3x3 mask', () => {
  const mask = testUtils.createMask([
    [1, 1, 1],
    [1, 0, 0],
    [0, 0, 1],
  ]);
  const roiMapManager = fromMask(mask);
  computeRois(roiMapManager);

  expect(roiMapManager.whiteRois).toHaveLength(2);
  expect(roiMapManager.blackRois).toHaveLength(1);

  expect(roiMapManager.whiteRois).toMatchSnapshot();
  expect(roiMapManager.blackRois).toMatchSnapshot();
});

test('test 2, waterShed for a grey image', () => {
  const image = testUtils.createGreyImage([
    [3, 3, 3, 3, 3, 3, 3, 3, 4, 4],
    [3, 3, 2, 2, 2, 3, 3, 3, 4, 4],
    [4, 3, 2, 1, 2, 2, 3, 3, 4, 4],
    [4, 3, 2, 2, 2, 2, 3, 3, 3, 4],
    [4, 4, 4, 3, 2, 3, 2, 3, 3, 4],
    [4, 4, 4, 3, 3, 3, 3, 1, 3, 3],
    [4, 3, 3, 3, 3, 3, 2, 2, 2, 3],
    [4, 4, 3, 3, 3, 3, 2, 2, 2, 2],
    [4, 4, 4, 4, 3, 2, 2, 2, 2, 3],
    [4, 4, 4, 4, 3, 3, 3, 3, 2, 3],
  ]);

  const roiMapManager = waterShed(image, { threshold: 2 / 255 });

  const result = new RoiMapManager({
    data: testUtils.getInt32Array(`
           0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
           0,  0, -1, -1, -1,  0,  0,  0,  0,  0,
           0,  0, -1, -1, -1, -1,  0,  0,  0,  0,
           0,  0, -1, -1, -1, -1,  0,  0,  0,  0,
           0,  0,  0,  0, -1,  0,  0,  0,  0,  0,
           0,  0,  0,  0,  0,  0,  0, -2,  0,  0,
           0,  0,  0,  0,  0,  0, -2, -2, -2,  0,
           0,  0,  0,  0,  0,  0, -2, -2, -2, -2,
           0,  0,  0,  0,  0, -2, -2, -2, -2,  0,
           0,  0,  0,  0,  0,  0,  0,  0, -2,  0
        `),
    nbPositive: 0,
    nbNegative: 2,
    width: 10,
    height: 10,
  });

  expect(roiMapManager).toStrictEqual(result);

  computeRois(roiMapManager);

  expect(roiMapManager.blackRois[0].origin).toStrictEqual({
    column: 2,
    row: 1,
  });
  expect(roiMapManager.blackRois[1].origin).toStrictEqual({
    column: 5,
    row: 5,
  });
});
