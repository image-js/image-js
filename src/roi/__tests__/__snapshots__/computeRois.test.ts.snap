// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3x3 mask 1`] = `
[
  {
    "centroid": {
      "column": 0.75,
      "row": 0.25,
    },
    "convexHull": {
      "perimeter": 9.23606797749979,
      "points": [
        {
          "column": 0,
          "row": 0,
        },
        {
          "column": 0,
          "row": 2,
        },
        {
          "column": 1,
          "row": 2,
        },
        {
          "column": 3,
          "row": 1,
        },
        {
          "column": 3,
          "row": 0,
        },
      ],
      "surface": 5,
    },
    "eqpc": 2.256758334191025,
    "feret": {
      "aspectRatio": 0.554700196225229,
      "maxDiameter": {
        "angle": -33.6900675259798,
        "calliperLines": [
          [
            {
              "column": -0.9230769230769232,
              "row": 0.6153846153846156,
            },
            {
              "column": 0.4615384615384618,
              "row": 2.6923076923076925,
            },
          ],
          [
            {
              "column": 2.0769230769230766,
              "row": -1.3846153846153848,
            },
            {
              "column": 3.4615384615384617,
              "row": 0.6923076923076923,
            },
          ],
        ],
        "length": 3.605551275463989,
        "points": [
          {
            "column": 0,
            "row": 2,
          },
          {
            "column": 3,
            "row": 0,
          },
        ],
      },
      "minDiameter": {
        "angle": 180,
        "calliperLines": [
          [
            {
              "column": 3,
              "row": 0,
            },
            {
              "column": -2.449293598294706e-16,
              "row": 3.67394039744206e-16,
            },
          ],
          [
            {
              "column": 3.0000000000000004,
              "row": 1.9999999999999996,
            },
            {
              "column": 0,
              "row": 2,
            },
          ],
        ],
        "length": 1.9999999999999996,
        "points": [
          {
            "column": 3,
            "row": 0,
          },
          {
            "column": 0,
            "row": 2,
          },
        ],
      },
    },
    "fillRatio": 1,
    "filledSurface": 4,
    "height": 2,
    "id": 1,
    "mbr": {
      "angle": -0,
      "aspectRatio": 0.6666666666666667,
      "height": 2,
      "perimeter": 10,
      "points": [
        {
          "column": 3,
          "row": -3.834758505292834e-17,
        },
        {
          "column": 3,
          "row": 2,
        },
        {
          "column": 3.2162452993532727e-16,
          "row": 2,
        },
        {
          "column": 4.440892098500626e-16,
          "row": -2.2204460492503128e-16,
        },
      ],
      "surface": 5.999999999999999,
      "width": 2.9999999999999996,
    },
    "origin": {
      "column": 0,
      "row": 0,
    },
    "ped": 2.0643291761037372,
    "perimeter": 6.485281374238571,
    "roundness": 0.39176601376466547,
    "solidity": 0.8,
    "sphericity": 1.0932163146821783,
    "surface": 4,
    "width": 3,
  },
  {
    "centroid": {
      "column": 2,
      "row": 2,
    },
    "convexHull": {
      "perimeter": 4,
      "points": [
        {
          "column": 0,
          "row": 0,
        },
        {
          "column": 0,
          "row": 1,
        },
        {
          "column": 1,
          "row": 1,
        },
        {
          "column": 1,
          "row": 0,
        },
      ],
      "surface": 1,
    },
    "eqpc": 1.1283791670955126,
    "feret": {
      "aspectRatio": 0.7071067811865475,
      "maxDiameter": {
        "angle": 45.00000000000001,
        "calliperLines": [
          [
            {
              "column": 0.5000000000000001,
              "row": -0.5,
            },
            {
              "column": -0.5,
              "row": 0.4999999999999999,
            },
          ],
          [
            {
              "column": 1.5,
              "row": 0.5,
            },
            {
              "column": 0.4999999999999998,
              "row": 1.5,
            },
          ],
        ],
        "length": 1.4142135623730951,
        "points": [
          {
            "column": 0,
            "row": 0,
          },
          {
            "column": 1,
            "row": 1,
          },
        ],
      },
      "minDiameter": {
        "angle": 90,
        "calliperLines": [
          [
            {
              "column": 0,
              "row": 0,
            },
            {
              "column": 6.123233995736766e-17,
              "row": 1,
            },
          ],
          [
            {
              "column": 1,
              "row": -6.123233995736766e-17,
            },
            {
              "column": 1,
              "row": 0.9999999999999999,
            },
          ],
        ],
        "length": 1,
        "points": [
          {
            "column": 0,
            "row": 0,
          },
          {
            "column": 1,
            "row": 0,
          },
        ],
      },
    },
    "fillRatio": 1,
    "filledSurface": 1,
    "height": 1,
    "id": 2,
    "mbr": {
      "angle": 0,
      "aspectRatio": 1,
      "height": 1,
      "perimeter": 4,
      "points": [
        {
          "column": 6.123233995736766e-17,
          "row": 1,
        },
        {
          "column": 0,
          "row": 0,
        },
        {
          "column": 1,
          "row": -6.123233995736766e-17,
        },
        {
          "column": 1,
          "row": 0.9999999999999999,
        },
      ],
      "surface": 1,
      "width": 1,
    },
    "origin": {
      "column": 2,
      "row": 2,
    },
    "ped": 0.4501581580785531,
    "perimeter": 1.4142135623730951,
    "roundness": 0.6366197723675812,
    "solidity": 1,
    "sphericity": 2.5066282746310002,
    "surface": 1,
    "width": 1,
  },
]
`;

exports[`3x3 mask 2`] = `
[
  {
    "centroid": {
      "column": 1,
      "row": 1.5,
    },
    "convexHull": {
      "perimeter": 8.82842712474619,
      "points": [
        {
          "column": 0,
          "row": 1,
        },
        {
          "column": 0,
          "row": 2,
        },
        {
          "column": 2,
          "row": 2,
        },
        {
          "column": 3,
          "row": 1,
        },
        {
          "column": 3,
          "row": 0,
        },
        {
          "column": 1,
          "row": 0,
        },
      ],
      "surface": 5,
    },
    "eqpc": 2.256758334191025,
    "feret": {
      "aspectRatio": 0.5547001962252291,
      "maxDiameter": {
        "angle": -33.6900675259798,
        "calliperLines": [
          [
            {
              "column": -0.6153846153846154,
              "row": 1.0769230769230773,
            },
            {
              "column": 0.6153846153846158,
              "row": 2.9230769230769234,
            },
          ],
          [
            {
              "column": 2.3846153846153846,
              "row": -0.9230769230769231,
            },
            {
              "column": 3.615384615384616,
              "row": 0.9230769230769231,
            },
          ],
        ],
        "length": 3.605551275463989,
        "points": [
          {
            "column": 0,
            "row": 2,
          },
          {
            "column": 3,
            "row": 0,
          },
        ],
      },
      "minDiameter": {
        "angle": 0,
        "calliperLines": [
          [
            {
              "column": 0,
              "row": 2,
            },
            {
              "column": 3,
              "row": 2,
            },
          ],
          [
            {
              "column": 0,
              "row": 0,
            },
            {
              "column": 3,
              "row": 0,
            },
          ],
        ],
        "length": 2,
        "points": [
          {
            "column": 0,
            "row": 2,
          },
          {
            "column": 3,
            "row": 0,
          },
        ],
      },
    },
    "fillRatio": 1,
    "filledSurface": 4,
    "height": 2,
    "id": -1,
    "mbr": {
      "angle": -0,
      "aspectRatio": 0.6666666666666667,
      "height": 2,
      "perimeter": 10,
      "points": [
        {
          "column": 3,
          "row": -3.834758505292834e-17,
        },
        {
          "column": 3,
          "row": 2,
        },
        {
          "column": 3.2162452993532727e-16,
          "row": 2,
        },
        {
          "column": 4.440892098500626e-16,
          "row": -2.2204460492503128e-16,
        },
      ],
      "surface": 5.999999999999999,
      "width": 2.9999999999999996,
    },
    "origin": {
      "column": 0,
      "row": 1,
    },
    "ped": 1.8006326323142123,
    "perimeter": 5.656854249492381,
    "roundness": 0.39176601376466547,
    "solidity": 0.8,
    "sphericity": 1.2533141373155001,
    "surface": 4,
    "width": 3,
  },
]
`;
